File: solutions/python/min-heap/06_size/min_heap.py
class MinHeap:
    def __init__(self) -> None:
        self._items = []

    def insert(self, value) -> None:
        self._items.append(value)
        self._bubble_up(len(self._items) - 1)

    def peek_min(self):
        if self.is_empty():
            return None
        return self._items[0]

    def extract_min(self):
        if self.is_empty():
            return None
        minimum = self._items[0]
        end = self._items.pop()
        if not self.is_empty():
            self._items[0] = end
            self._bubble_down(0)
        return minimum

    def size(self) -> int:
        return len(self._items)

    def is_empty(self) -> bool:
        return len(self._items) == 0

    def _bubble_up(self, index: int) -> None:
        current = index
        while current > 0:
            parent = (current - 1) // 2
            if self._items[parent] <= self._items[current]:
                break
            self._items[parent], self._items[current] = (
                self._items[current],
                self._items[parent],
            )
            current = parent

    def _bubble_down(self, index: int) -> None:
        length = len(self._items)
        current = index
        while True:
            left = current * 2 + 1
            right = current * 2 + 2
            smallest = current

            if left < length and self._items[left] < self._items[smallest]:
                smallest = left
            if right < length and self._items[right] < self._items[smallest]:
                smallest = right
            if smallest == current:
                break
            self._items[current], self._items[smallest] = (
                self._items[smallest],
                self._items[current],
            )
            current = smallest

File: solutions/javascript/min-heap/05-heapify-down/minHeap.js
class MinHeap {
  constructor() {
    this._items = [];
  }

  insert(value) {
    this._items.push(value);
    this._bubbleUp(this._items.length - 1);
  }

  peekMin() {
    if (this._items.length === 0) {
      return undefined;
    }
    return this._items[0];
  }

  extractMin() {
    if (this._items.length === 0) {
      return undefined;
    }
    const min = this._items[0];
    const end = this._items.pop();
    if (this._items.length > 0) {
      this._items[0] = end;
      this._bubbleDown(0);
    }
    return min;
  }

  _bubbleUp(index) {
    let current = index;
    while (current > 0) {
      const parent = Math.floor((current - 1) / 2);
      if (this._items[parent] <= this._items[current]) {
        break;
      }
      this._swap(parent, current);
      current = parent;
    }
  }

  _bubbleDown(index) {
    let current = index;
    const length = this._items.length;
    while (true) {
      const left = current * 2 + 1;
      const right = current * 2 + 2;
      let smallest = current;

      if (left < length && this._items[left] < this._items[smallest]) {
        smallest = left;
      }
      if (right < length && this._items[right] < this._items[smallest]) {
        smallest = right;
      }
      if (smallest === current) {
        break;
      }
      this._swap(current, smallest);
      current = smallest;
    }
  }

  _swap(i, j) {
    const temp = this._items[i];
    this._items[i] = this._items[j];
    this._items[j] = temp;
  }
}

module.exports = MinHeap;

